### PR TITLE
Leaderboard button in /xp rank!!!!!!

### LINKED ISSUE #345: put a leaderboard button in /xp rank
jump to the user ofc

### PR DESCRIPTION
**Pull request by Platformer444**
resolves #345.

Dont scream at me if it isnt what you want.
Shut up grammarly

### CODE REVIEW

File: modules/xp/index.ts [hunk 1]
```diff
@@ -135,7 +87,62 @@ defineButton("xp", async (interaction, userId = "") => {
...
+defineButton("viewLeaderboard", async (interaction, userId) => {
+	await XPLeaderboard(interaction, (await config.guild.members.fetch(userId)).user);
+});
+
 if (constants.canvasEnabled) {
 	const { default: weeklyXpGraph } = await import("./graph.js");
 	defineSelect("weeklyXpGraph", weeklyXpGraph);
 }
+
+export async function XPLeaderboard(interaction: ChatInputCommandInteraction<"raw" | "cached"> | ButtonInteraction, user: User | undefined) {
```
Comment (RedGuy12) [inline modules/xp/index.ts:99]: Rename `XPLeaderboard` to `top`

File: modules/xp/index.ts [hunk 2]
```diff
@@ -135,7 +87,62 @@ defineButton("xp", async (interaction, userId = "") => {
 	await getUserRank(interaction, await client.users.fetch(userId));
 });
 
+defineButton("viewLeaderboard", async (interaction, userId) => {
+	await XPLeaderboard(interaction, (await config.guild.members.fetch(userId)).user);
```
Comment (RedGuy12) [inline modules/xp/index.ts:91]: Use `client.users.fetch`

File: modules/xp/index.ts [hunk 3]
```diff
@@ -78,55 +78,7 @@ defineSubcommands(
...
-											label: "Toggle Mentions",
-											style: ButtonStyle.Success,
-										},
-								  ]
-								: undefined;
-						},
-						customComponentLocation: "below",
-					},
-				);
+				await XPLeaderboard(interaction, options.options.user as User);
```
Comment (RedGuy12) [inline modules/xp/index.ts:81]: Don't do `as User`, the function already supports GuildMembers, just change the signature

**Review by RedGuy12 — COMMENTED** [review]
This looks great, thanks! Just a few nitpicks

File: modules/xp/index.ts (hunk 3 above)
Comment (Platformer444) [reply modules/xp/index.ts]: For some reason TypeScript was screaming at me when I didn't do `as User`

Comment (Platformer444) [reply modules/xp/index.ts]: Oh wait now I remember. I just have to add `.user`

File: modules/xp/index.ts [hunk 4]
```diff
@@ -78,55 +78,7 @@ defineSubcommands(
...
-											label: "Toggle Mentions",
-											style: ButtonStyle.Success,
-										},
-								  ]
-								: undefined;
-						},
-						customComponentLocation: "below",
-					},
-				);
+				await top(interaction, options.options.user.user);
```
Comment (RedGuy12) [inline modules/xp/index.ts:81]:
```suggestion
				await top(interaction, options.options.user);
```

File: modules/xp/index.ts [hunk 5]
```diff
@@ -135,7 +87,62 @@ defineButton("xp", async (interaction, userId = "") => {
...
+defineButton("viewLeaderboard", async (interaction, userId) => {
+	await top(interaction, await client.users.fetch(userId));
+});
+
 if (constants.canvasEnabled) {
 	const { default: weeklyXpGraph } = await import("./graph.js");
 	defineSelect("weeklyXpGraph", weeklyXpGraph);
 }
+
+export async function top(interaction: ChatInputCommandInteraction<"raw" | "cached"> | ButtonInteraction, user: User | undefined) {
```
Comment (RedGuy12) [inline modules/xp/index.ts:99]:
```suggestion
export async function top(interaction: ChatInputCommandInteraction<"raw" | "cached"> | ButtonInteraction, user?: User | GuildMember) {
```

File: modules/xp/index.ts (hunk 3 above)
Comment (RedGuy12) [reply modules/xp/index.ts]: no, it was because you hadn't changed the function signature

**Pull request merged**